### PR TITLE
Allow trailing optional args in array >{...} syntax

### PR DESCRIPTION
**Pull request by josephwright**
**READ ME FIRST**: Please understand that in most cases we will not be able to merge a pull request because there are a lot of internal activities needed when updating the LaTeX2e sources. If you have a code suggestion please discuss it with the team first.

*Pull requests in this repository are intended for LaTeX Team members only.*

# Internal housekeeping

## Status of pull request

- Feedback wanted 
- Under development
- [x] Ready to merge

## Checklist of required changes before merge will be approved
- [ ] Test file(s) added
- [x] Version and date string updated in changed source files
- [x] Relevant `\changes` entries in source included
- [x] Relevant `changes.txt` updated
- [ ] Rollback provided (if necessary)?
- [ ] `ltnewsX.tex` (and/or `latexchanges.tex`) updated

### CODE REVIEW

**Comment by josephwright** [comment]
No new tests as the ones that show things up are now actually using the relevant code!

**Comment by josephwright** [comment]
Also, no update to the `ltnews` entry as that already covers things, really.

File: required/tools/array.dtx [hunk 1]
```diff
@@ -1388,11 +1389,29 @@ Bug reports can be opened (category \texttt{#1}) at\\%
...
 %    =\@tempcnta= has saved the value $=\count@= - 1$.
 %
-%    We end with \cs{relax} to stop any parsing for optional
-%    arguments out of \verb=>{...}= at this point. 
+%    To keep \TeX{} happy if there is a look ahead in the tabular preamble
+%    which uses the Appendix~D trick (for example anything with a trailing
+%    optional argument defined by \pkg{ltcmd}), we wrap everything here in
+%    a protected version of \cs{@firstofone}. \TeX{} otherwise can get
+%    confused about the value of the master counter, and we get some
+%    strange errors. (Quite possibly the underlying issue is a \TeX{}
```
Comment (FrankMittelbach) [inline required/tools/array.dtx:1397]: We might confess in a footnote that we don't quite understand what's happening right now ... maybe somebody sets us straight or finds the bug if there is one

File: required/tools/array.dtx [hunk 2]
```diff
@@ -1401,13 +1420,13 @@ Bug reports can be opened (category \texttt{#1}) at\\%
 %    =\@sharp= which is temporarily occupied with
 %    =\relax= during the build-up of the preamble.
 %    To remove unwanted spaces before and after the column text, we set
-%    an =\ignorespaces=  in front and a =\unskip= afterwards.
+%    an =\ignorespaces=  in front (see above) and a =\unskip= afterwards.
```
Comment (FrankMittelbach) [inline required/tools/array.dtx:1423]: ... but only if the cell isn't a math cell.

File: required/latex-lab/config-math-luatex.lua [hunk 3]
```diff
@@ -5,12 +5,6 @@ checkengines = {"luatex"}
 checksearch  = true
 testfiledir  = "testfiles-math-luatex"
 
-checkdeps =
-  {
-    maindir .. "/base",
-    maindir .. "/required/amsmath"
-  }
-
```
Comment (FrankMittelbach) [inline required/latex-lab/config-math-luatex.lua:13]: not quite sure I understand why this means it is using the tools dir as a dependency.

**Review by FrankMittelbach — APPROVED** [review]
approved, with some minor comments and a question

File: required/latex-lab/config-math-luatex.lua (hunk 3 above)
Comment (u-fischer) [reply required/latex-lab/config-math-luatex.lua:13]: the main build.lua sets checkdeps to a larger set (which includes tools).

Comment (josephwright) [reply required/latex-lab/config-math-luatex.lua:13]: @FrankMittelbach  What @u-fischer  said - we had a config-dependent override which didn't really make sense

File: required/tools/array.dtx (hunk 2 above)
Comment (josephwright) [reply required/tools/array.dtx:1423]: I've not changed anything here other than the code is now above the comment not below - do I need to change more?

File: required/tools/array.dtx (hunk 1 above)
Comment (josephwright) [reply required/tools/array.dtx:1397]: I've added something

File: required/latex-lab/config-math-luatex.lua (hunk 3 above)
Comment (FrankMittelbach) [reply required/latex-lab/config-math-luatex.lua:13]: well I'm not sure it makes sense to have this hierachical. Seems to me it is better to set dependencies always for the test dir you work with, but fine, something to remember.

File: required/tools/array.dtx (hunk 2 above)
Comment (FrankMittelbach) [reply required/tools/array.dtx:1428]: I know, I just meant it makes sense to say why \textonly@unskip ... we (or I) neglected that when it got introduced

File: required/latex-lab/config-math-luatex.lua (hunk 3 above)
Comment (josephwright) [reply required/latex-lab/config-math-luatex.lua:13]: We have a standard set of deps for the entire repo - the main reason for altering is where you'd get a circular ref

**Comment by FrankMittelbach** [comment]
well, something is badly breaking in the hook tests ...

**Comment by josephwright** [comment]
> well, something is badly breaking in the hook tests ...

Er, that was a mess-up between branches, which I've sorted with a force-pull (I was trying something unrelated and got the wrong source committed) - try now

**Comment by josephwright** [comment]
Just waiting for test suite to finish, then I'll merge

**Review by davidcarlisle — APPROVED** [review]

**Comment by muzimuzhi** [comment]
(just adding a link to #1468)

File: required/tools/array.dtx [hunk 4]
```diff
@@ -1432,10 +1458,15 @@ Bug reports can be opened (category \texttt{#1}) at\\%
 %    \begin{macrocode}
 \protected\def\textonly@unskip{\ifhmode\unskip\fi}
 %    \end{macrocode}
+% \changes{v2.6g}{2024/10/12}{Further work to support optional args in preamble (gh/1468)}
+%  We need an engine-protected function that is just \cs{@firstofone}:
+%    \begin{macrocode}
+\protected\def\@protected@firstofone#1{#1}
```
Comment (Skillmon) [inline required/tools/array.dtx:1464]: Should be `\long` (though the usage here isn't requiring that).

Comment (muzimuzhi) [reply required/tools/array.dtx:1464]: Added, see 7db95fb6 (Add a missing \long, 2024-10-12) by Joseph.

**Pull request merged**